compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -168,6 +168,23 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
 
             ty::Infer(_) | ty::Bound(_, _) => bug!("unexpected type `{self_ty}`"),
 
+            // Generators have one special built-in candidate, `Unpin`, which
+            // takes precedence over the structural auto trait candidate being
+            // assembled.
```

Comment:
conceptually what's happening here is that we have an *explicit* builtin impl of unpin for generators. the idea for auto traits is

```
fn consider_auto_trait_candidate(..) {
    if outmost_type_constructor_has_explicit_impl() {
        return Err(NoSolution);
    }

    add_auto_trait_candidate
}

fn consider_builtin_unpin_candidate(..) {
    // adds the candidate for generators.
}
```

where `outmost_type_constructor_has_explicit_impl` considers both user written and builtin impls. I am a bit worried that going forward we may try to add a builtin candidate for yet another auto trait but do so outside of `consider_auto_trait_candidate`. 

I don't think it really makes sense to actually have a `outmost_type_constructor_has_explicit_impl` and `consider_builtin_unpin_candidate` split. What do you think about

- move this whole match into a separate function, something like `disable_auto_trait_candidate_due_to_explicit_impl` which returns `Option<QueryResult<'tcx>>` with a comment mentioning that this function considers builtin candidates for auto traits
- add a comment in `assemble_builtin_impl_candidates` that for auto traits, builtin impls have to be added to `disable_auto_trait_candidate_due_to_explicit_impl` instead

Reply:
Hmm.. took a stab at that. Let me know what you think.